custom_components/kia_uvo/number.py

```diff
@@ -0,0 +1,119 @@
+
+from .const import DOMAIN
+from .coordinator import HyundaiKiaConnectDataUpdateCoordinator
+from .entity import HyundaiKiaConnectEntity
+
+_LOGGER = logging.getLogger(__name__)
+
+AC_CHARGING_LIMIT_KEY = "_ac_charging_limit"
+DC_CHARGING_LIMIT_KEY = "_dc_charging_limit"
+
+CHARGING_LIMIT_DESCRIPTIONS: Final[tuple[NumberEntityDescription, ...]] = (
+    NumberEntityDescription(
+        key=AC_CHARGING_LIMIT_KEY,
+        name="AC Charging Limit",
+        icon="mdi:ev-plug-type2",
+        min_value=50,
+        max_value=100,
+        step=10,
+    ),
+    NumberEntityDescription(
+        key=DC_CHARGING_LIMIT_KEY,
+        name="DC Charging Limit",
+        icon="mdi:ev-plug-ccs2",
+        min_value=50,
+        max_value=100,
+        step=10,
+    ),
+)
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    coordinator = hass.data[DOMAIN][config_entry.unique_id]
+    entities = []
+    for vehicle_id in coordinator.vehicle_manager.vehicles.keys():
+        vehicle: Vehicle = coordinator.vehicle_manager.vehicles[vehicle_id]
+        for descr in CHARGING_LIMIT_DESCRIPTIONS:
+            entities.append(HyundaiKiaChargingLimitNumber(coordinator, descr, vehicle))
+
+    async_add_entities(entities)
+    return True
+
+
+class HyundaiKiaChargingLimitNumber(NumberEntity, HyundaiKiaConnectEntity):
+    _vehicle_manager: VehicleManager
+    _vehicle: Vehicle
+
+    def __init__(
+        self,
+        coordinator: HyundaiKiaConnectDataUpdateCoordinator,
+        description: NumberEntityDescription,
+        vehicle: Vehicle,
+    ) -> None:
+        HyundaiKiaConnectEntity.__init__(self, coordinator, vehicle)
+        self._attr_unique_id = f"{DOMAIN}_{vehicle.id}_{description.key}"
+        self.entity_description = description
+        self._vehicle_manager = coordinator.vehicle_manager
+        self._vehicle = vehicle
+
+    @property
+    def value(self) -> float | None:
+        """Return the entity value to represent the entity state."""
+        if self.entity_description.key == AC_CHARGING_LIMIT_KEY:
+            return self._vehicle.ev_charge_limits.ac
+        else:
+            return self._vehicle.ev_charge_limits.dc
+
+    async def async_set_value(self, value: float) -> None:
+        """Set new charging limit."""
+        # force refresh of state so that we can get the value for the other charging limit
+        # since we have to set both limits as compound API call.
+        current_limits = await self.hass.async_add_executor_job(
+            self._vehicle_manager.api.get_charge_limits,
+            self._vehicle_manager.token,
+            self.vehicle,
+        )
+
+        # don't do anything for null change
+        if (
+            self.entity_description.key == AC_CHARGING_LIMIT_KEY
+            and current_limits.ac == int(value)
+        ):
+            return
```

Comment:
here is a bug - when setting the limits on a different channel (e.g. bluelink app) and then pulling the slider in HA to the same target percentage, it stops here and the UI will stay at the old displayValue - until the configured refresh cycle finally hits. will fix soon

Reply:
@jwefers were you ever able to get to fixing this ? I tried it and was able to set the charge limits and couldn't really recreate the issue. Thanks.

Reply:
No, i did not find the time since March to work on this :D the bug is obviously still present - eventually i need to rethink myself into it. I think what needs to be done here is to consider the internally cached state with what the API returns AND the desired number.